src/session.rs

```diff
@@ -656,7 +687,13 @@ mod impl_socks5 {
 
         fn advance(&mut self, input: &[u8]) -> Result<Vec<u8>, Self::Error> { self.advance(input) }
 
-        fn artifact(&self) -> Option<Self::Artifact> { Some(()) }
+        fn artifact(&self) -> Option<Self::Artifact> {
+            match self {
+                Socks5::Initial(addr, false) if !addr.requires_proxy() => Some(addr.clone()),
```

Comment:
This line is the actual bugfix: we return artifact signifying successful handshake if the proxy is not used and not needed. Thus, we do not run a empty handshake and successfully activate handshake in up-level session.